app/src/main/java/com/google/android/sambadocumentsprovider/encryption/EncryptionManager.java

```diff
@@ -0,0 +1,177 @@
+    try {
+      Cipher cipher = Cipher.getInstance(AES_CIPHER);
+      cipher.init(
+              Cipher.ENCRYPT_MODE, mKey.getKey(),
+              new GCMParameterSpec(GCM_TAG_LENGTH, mKey.getIv()));
+
+      byte[] encrypted = cipher.doFinal(data.getBytes(Charset.forName(DEFAULT_CHARSET)));
+
+      return Base64.encodeToString(encrypted, Base64.DEFAULT);
+    } catch (Exception e) {
+      throw new EncryptionException("Failed to encrypt data: ", e);
+    }
+  }
+
+  public String decrypt(String data) throws EncryptionException {
+    try {
+      Cipher cipher = Cipher.getInstance(AES_CIPHER);
+      cipher.init(
+              Cipher.DECRYPT_MODE, mKey.getKey(),
+              new GCMParameterSpec(GCM_TAG_LENGTH, mKey.getIv()));
+
+      byte[] decrypted = cipher.doFinal(Base64.decode(data, Base64.DEFAULT));
+      return new String(decrypted, Charset.forName(DEFAULT_CHARSET));
+    } catch (Exception e) {
+      throw new EncryptionException("Failed to decrypt data: ", e);
+    }
+  }
+
+  private void loadKey() throws EncryptionException {
+    try {
+      SecretKey key = (SecretKey) mStore.getKey(KEY_ALIAS, null);
+      if (key != null) {
+        mKey = new EncryptionKey(key, loadIv());
+        return;
+      }
+
+      KeyGenerator keyGen = KeyGenerator.getInstance(
+              KeyProperties.KEY_ALGORITHM_AES, KEYSTORE_PROVIDER);
+
+      KeyGenParameterSpec spec = new KeyGenParameterSpec.Builder(
+              KEY_ALIAS, KeyProperties.PURPOSE_ENCRYPT | KeyProperties.PURPOSE_DECRYPT)
+              .setBlockModes(KeyProperties.BLOCK_MODE_GCM)
+              .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_NONE)
+              .setRandomizedEncryptionRequired(false)
+              .build();
+      keyGen.init(spec);
+
+      key = keyGen.generateKey();
+      byte[] iv = generateIv();
+
+      mKey = new EncryptionKey(key, iv);
+      saveIv(iv);
+    } catch (Exception e) {
+      throw new EncryptionException("Failed to register key: ", e);
+    }
+  }
+
+  private void loadKeyStore() throws KeyStoreException, CertificateException, NoSuchAlgorithmException, IOException {
+    mStore = KeyStore.getInstance(KEYSTORE_PROVIDER);
+    mStore.load(null);
+  }
+
+  private static byte[] generateIv() {
+    byte[] iv = new byte[IV_LENGTH];
+
+    mRandom.nextBytes(iv);
+
+    return iv;
+  }
+
+  private byte[] loadIv() {
+    String data = mPref.getString(KEY_ALIAS, null);
+    if (data == null) {
+      byte[] iv = generateIv();
+      saveIv(iv);
+      return iv;
+    }
+
+    return Base64.decode(data, Base64.DEFAULT);
+  }
+
+  private void saveIv(byte[] iv) {
+    String data = Base64.encodeToString(iv, Base64.DEFAULT);
+    mPref.edit().putString(KEY_ALIAS, data).apply();
+  }
+
+  private static class EncryptionKey {
+    private final SecretKey mKey;
+    private final byte[] mIv;
+
+    EncryptionKey(SecretKey key, byte[] iv) {
+      mKey = key;
+      mIv = iv;
+    }
+
+    SecretKey getKey() {
```

Comment:
nit: private getters or put it in a separate file as you already have a package.